### PR TITLE
Draft: Add strong params to JSON deserialization

### PR DESCRIPTION
**Pull request by dukenguyenxyz**

### CODE REVIEW

File: src/clear/model/json_deserialize.cr [hunk 1]
```diff
@@ -24,25 +24,26 @@ macro columns_to_instance_vars
...
+      assign_columns(model, permit, mass_assignment)
     end
 
     macro finished
       # Assign properties to the model inputted with object's instance variables
-      protected def assign_columns(model)
+      protected def assign_columns(model, permit, mass_assignment)
         {% for name, settings in COLUMNS %}
-          if self.{{name.id}}_present?
+          if self.{{name.id}}_present? && ((mass_assignment && permit.size == 0) ? true : permit.includes?({{name}}))
```
Comment (caspiano) [inline src/clear/model/json_deserialize.cr]:
```suggestion
      protected def assign_columns(model, permit : Enumerable(String | Symbol), mass_assignment : Bool)
        {% for name, settings in COLUMNS %}
          {% name_string = name.stringify %}
          {% name_symbol = name.symbolize %}
          if self.{{name.id}}_present? && mass_assignment && permit.any? { |p| p == {{name_string.id}}) || p == {{name_symbol.id}} } 
```

File: src/clear/model/json_deserialize.cr [hunk 2]
```diff
@@ -60,16 +61,16 @@ macro columns_to_instance_vars
   # Create a new empty model and fill the columns from json
   #
   # Returns the new model
-  def self.from_json(string_or_io : String | IO)
-    Assigner.from_json(string_or_io).create
+  def self.from_json(string_or_io : String | IO, permit : Array(String) = [] of String, mass_assignment : Bool = true)
```
Comment (caspiano) [inline src/clear/model/json_deserialize.cr]:
```suggestion
  def self.from_json(string_or_io : String | IO, permit : Enumerable(String | Symbol) = [] of String | Symbol, mass_assignment : Bool = true)
```

File: src/clear/model/json_deserialize.cr [hunk 3]
```diff
@@ -60,16 +61,16 @@ macro columns_to_instance_vars
...
+    Assigner.from_json(string_or_io).create(permit, mass_assignment)
   end
 
   # Create a new model from json and save it. Returns the model.
   #
   # The model may not be saved due to validation failure;
   # check the returned model `errors?` and `persisted?` flags.
-  def self.create_from_json(string_or_io : String | IO)
-    mdl = self.from_json(string_or_io)
+  def self.create_from_json(string_or_io : String | IO, permit : Array(String) = [] of String, mass_assignment : Bool = true)
```
Comment (caspiano) [inline src/clear/model/json_deserialize.cr]:
```suggestion
  def self.create_from_json(string_or_io : String | IO, permit : Enumerable(String | Symbol) = [] of String | Symbol, mass_assignment : Bool = true)
```

File: src/clear/model/json_deserialize.cr [hunk 4]
```diff
@@ -78,24 +79,24 @@ macro columns_to_instance_vars
   #
   # Returns the newly inserted model
   # Raises an exception if validation failed during the saving process.
-  def self.create_from_json!(string_or_io : String | IO)
-    self.from_json(string_or_io).save!
+  def self.create_from_json!(string_or_io : String | IO, permit : Array(String) = [] of String, mass_assignment : Bool = true)
```
Comment (caspiano) [inline src/clear/model/json_deserialize.cr]:
```suggestion
  def self.create_from_json!(string_or_io : String | IO, permit : Enumerable(String | Symbol) = [] of String | Symbol, mass_assignment : Bool = true)
```

File: src/clear/model/json_deserialize.cr [hunk 5]
```diff
@@ -78,24 +79,24 @@ macro columns_to_instance_vars
...
+    self.from_json(string_or_io, permit, mass_assignment).save!
   end
 
   # Set the fields from json passed as argument
-  def set_from_json(string_or_io : String | IO)
-    Assigner.from_json(string_or_io).update(self)
+  def set_from_json(string_or_io : String | IO, permit : Array(String) = [] of String, mass_assignment : Bool = true)
```
Comment (caspiano) [inline src/clear/model/json_deserialize.cr]:
```suggestion
  def set_from_json(string_or_io : String | IO, permit : Enumerable(String | Symbol) = [] of String | Symbol, mass_assignment : Bool = true)

File: src/clear/model/json_deserialize.cr [hunk 6]
```diff
@@ -78,24 +79,24 @@ macro columns_to_instance_vars
...
+    Assigner.from_json(string_or_io).update(self, permit, mass_assignment)
   end
 
   # Set the fields from json passed as argument and call `save` on the object
-  def update_from_json(string_or_io : String | IO)
-    mdl = set_from_json(string_or_io)
+  def update_from_json(string_or_io : String | IO, permit : Array(String) = [] of String, mass_assignment : Bool = true)
```
Comment (caspiano) [inline src/clear/model/json_deserialize.cr]:
```suggestion
  def update_from_json(string_or_io : String | IO, permit : Enumerable(String | Symbol) = [] of String | Symbol, mass_assignment : Bool = true)
```

File: src/clear/model/json_deserialize.cr [hunk 7]
```diff
@@ -78,24 +79,24 @@ macro columns_to_instance_vars
...
+    mdl = set_from_json(string_or_io, permit, mass_assignment)
     mdl.save
     mdl
   end
 
   # Set the fields from json passed as argument and call `save!` on the object
-  def update_from_json!(string_or_io : String | IO)
-    set_from_json(string_or_io).save!
+  def update_from_json!(string_or_io : String | IO, permit : Array(String) = [] of String, mass_assignment : Bool = true)
```
Comment (caspiano) [inline src/clear/model/json_deserialize.cr]:
```suggestion
  def update_from_json!(string_or_io : String | IO, permit : Enumerable(String | Symbol) = [] of String | Symbol, mass_assignment : Bool = true)
```

File: src/clear/model/json_deserialize.cr (hunk 2 above)
Comment (dukenguyenxyz) [reply src/clear/model/json_deserialize.cr]: What is the difference between using Array / Enumerable. When should I use which?

File: src/clear/model/json_deserialize.cr [hunk 8]
```diff
@@ -24,25 +24,26 @@ macro columns_to_instance_vars
...
+      assign_columns(model, permit, mass_assignment)
     end
 
     macro finished
       # Assign properties to the model inputted with object's instance variables
-      protected def assign_columns(model)
+      protected def assign_columns(model, permit : Enumerable(String | Symbol), mass_assignment : Bool)
         {% for name, settings in COLUMNS %}
-          if self.{{name.id}}_present?
+          if self.{{name.id}}_present? && ((mass_assignment && permit.size == 0) ? true : (permit.any? { |p| p == {{name.stringify.id}} || p == {{name.id.symbolize}} }))
```
Comment (caspiano) [inline src/clear/model/json_deserialize.cr:46]: this condition is quite confusing, and can be made simpler
not really sure what the purpose of the `mass_assignment` variable is here
```suggestion
           if self.{{name.id}}_present? && (!mass_assignment || permit.any? { |p| p == {{name.id.stringify}} || p == {{name.id.symbolize}} })
```

**Pull request closed without merging**